File: messages-view/ios/Sources/MessagesView/MessageTheme.swift
import UIKit

/// Visual configuration for the renderer. Colours/sizes are supplied here rather than pulled
/// from a global, so the library stays a pure, importable component. Palette matches the
/// web/Android renderers so screenshots line up.
public struct MessageTheme {
    public var textColor: UIColor
    public var titleColor: UIColor
    public var subtitleColor: UIColor
    public var descriptionColor: UIColor
    public var overlayTextColor: UIColor
    public var linkColor: UIColor
    public var redactionColor: UIColor
    public var bubbleColor: UIColor
    public var dividerColor: UIColor

    public var defaultTextSize: CGFloat
    public var titleTextSize: CGFloat
    public var subtitleTextSize: CGFloat
    public var descriptionTextSize: CGFloat

    public var bubbleRadius: CGFloat
    public var bubbleMaxWidth: CGFloat

    public var buttonTextColor: UIColor
    public var ctaTextColor: UIColor
    public var ctaBackgroundColor: UIColor

    private static let white70 = UIColor(white: 1, alpha: 0.7)
    private static let white60 = UIColor(white: 1, alpha: 0.6)
    private static let white50 = UIColor(white: 1, alpha: 0.5)
    private static let white15 = UIColor(white: 1, alpha: 0.15)
    private static let link = UIColor(rgb: 0x8E84FA)
    private static let redaction = UIColor(white: 0, alpha: 0.85)

    private static func base(text: UIColor, title: UIColor, subtitle: UIColor, description: UIColor, bubble: UIColor) -> MessageTheme {
        MessageTheme(
            textColor: text,
            titleColor: title,
            subtitleColor: subtitle,
            descriptionColor: description,
            overlayTextColor: .white,
            linkColor: link,
            redactionColor: redaction,
            bubbleColor: bubble,
            dividerColor: white15,
            defaultTextSize: 16,
            titleTextSize: 18,
            subtitleTextSize: 15,
            descriptionTextSize: 13,
            bubbleRadius: 18,
            bubbleMaxWidth: 260,
            buttonTextColor: link,
            ctaTextColor: .white,
            ctaBackgroundColor: link
        )
    }

    /// Message received from someone else: neutral dark bubble.
    public static let incoming = base(text: .white, title: .white, subtitle: white70, description: white50, bubble: UIColor(rgb: 0x262532))

    /// Our own message: accent bubble.
    public static let outgoing = base(text: .white, title: .white, subtitle: white70, description: white50, bubble: UIColor(rgb: 0x7924FF))

    /// Compact preview used in a room/conversation list row.
    public static let preview = base(text: white60, title: white60, subtitle: white60, description: white60, bubble: .clear)
}

extension UIColor {
    convenience init(rgb: Int) {
        self.init(
            red: CGFloat((rgb >> 16) & 0xFF) / 255,
            green: CGFloat((rgb >> 8) & 0xFF) / 255,
            blue: CGFloat(rgb & 0xFF) / 255,
            alpha: 1
        )
    }

    /// Packed ARGB int (as carried in ImageReference.preview_color).
    convenience init(argb: Int32) {
        let value = UInt32(bitPattern: argb)
        self.init(
            red: CGFloat((value >> 16) & 0xFF) / 255,
            green: CGFloat((value >> 8) & 0xFF) / 255,
            blue: CGFloat(value & 0xFF) / 255,
            alpha: value > 0xFFFFFF ? CGFloat((value >> 24) & 0xFF) / 255 : 1
        )
    }
}

File: messages-view/android/src/main/kotlin/com/latenighthack/social/messages/view/MessageTheme.kt
package com.latenighthack.social.messages.view

import android.content.Context
import android.graphics.Color

/**
 * Visual configuration for the renderer. Ported from the reference client's `MessageTheme`,
 * with colours/margins supplied here rather than pulled from a global theming singleton so the
 * library stays pure and importable. Text sizes are in sp; margins/radius are resolved to px.
 */
data class MessageTheme(
    val defaultTextSize: Float,
    val titleTextSize: Float,
    val subtitleTextSize: Float,
    val descriptionTextSize: Float,
    val textHorizontalMargin: Int,
    val textVerticalMargin: Int,
    val textBottomMargin: Int,
    val textColor: Int,
    val linkTextColor: Int,
    val titleTextColor: Int,
    val subtitleTextColor: Int,
    val descriptionTextColor: Int,
    val bubbleBackgroundColor: Int,
    val overlayTextColor: Int,
    val overlayTitleTextColor: Int,
    val overlaySubtitleTextColor: Int,
    val overlayDescriptionTextColor: Int,
    val dividerColor: Int,
    val redactionColor: Int,
    val bubbleRadius: Float,
    val buttonTextColor: Int,
    val ctaTextColor: Int,
    val ctaBackgroundColor: Int,
    val isImmersive: Boolean = false,
) {
    companion object {
        // Palette shared with the web/iOS renderers so screenshots line up.
        private const val WHITE = 0xFFFFFFFF.toInt()
        private const val WHITE70 = 0xB3FFFFFF.toInt()
        private const val WHITE60 = 0x99FFFFFF.toInt()
        private const val WHITE50 = 0x80FFFFFF.toInt()
        private const val WHITE15 = 0x26FFFFFF.toInt()
        private const val LINK = 0xFF8E84FA.toInt()
        private const val REDACTION = 0xD9000000.toInt()
        private const val INCOMING_BUBBLE = 0xFF262532.toInt()
        private const val OUTGOING_BUBBLE = 0xFF7924FF.toInt()

        private fun base(
            context: Context,
            textColor: Int,
            titleColor: Int,
            subtitleColor: Int,
            descriptionColor: Int,
            bubble: Int,
        ): MessageTheme = MessageTheme(
            defaultTextSize = 16f,
            titleTextSize = 18f,
            subtitleTextSize = 15f,
            descriptionTextSize = 13f,
            textHorizontalMargin = context.dpToPxInt(12f),
            textVerticalMargin = context.dpToPxInt(6f),
            textBottomMargin = context.dpToPxInt(8f),
            textColor = textColor,
            linkTextColor = LINK,
            titleTextColor = titleColor,
            subtitleTextColor = subtitleColor,
            descriptionTextColor = descriptionColor,
            bubbleBackgroundColor = bubble,
            overlayTextColor = WHITE,
            overlayTitleTextColor = WHITE,
            overlaySubtitleTextColor = WHITE70,
            overlayDescriptionTextColor = WHITE50,
            dividerColor = WHITE15,
            redactionColor = REDACTION,
            bubbleRadius = context.dpToPx(18f),
            buttonTextColor = LINK,
            ctaTextColor = WHITE,
            ctaBackgroundColor = LINK,
        )

        /** Message received from someone else: neutral dark bubble. */
        fun incoming(context: Context): MessageTheme =
            base(context, WHITE, WHITE, WHITE70, WHITE50, INCOMING_BUBBLE)

        /** Our own message: accent bubble. */
        fun outgoing(context: Context): MessageTheme =
            base(context, WHITE, WHITE, WHITE70, WHITE50, OUTGOING_BUBBLE)

        /** Compact preview used in a room/conversation list row. */
        fun preview(context: Context): MessageTheme =
            base(context, WHITE60, WHITE60, WHITE60, WHITE60, Color.TRANSPARENT)
    }
}
